pkg/appmeshinject/sidecar.go

```diff
@@ -0,0 +1,158 @@
+package appmeshinject
+
+import (
+	"bufio"
+	"bytes"
+	"text/template"
+)
+
+const envoyContainerTemplate = `
+{
+  "name": "envoy",
+  "image": "{{ .ContainerImage }}",
+  "securityContext": {
+    "runAsUser": 1337
+  },
+  "ports": [
+    {
+      "containerPort": 9901,
+      "name": "stats",
+      "protocol": "TCP"
+    }
+  ],
+  "env": [
+    {
+      "name": "APPMESH_VIRTUAL_NODE_NAME",
+      "value": "mesh/{{ .MeshName }}/virtualNode/{{ .VirtualNodeName }}"
+    },
+    {
+      "name": "APPMESH_PREVIEW",
+      "value": "{{ .Preview }}"
+    },
+    {
+      "name": "ENVOY_LOG_LEVEL",
+      "value": "{{ .LogLevel }}"
+    }{{ if or .EnableJaegerTracing .EnableDatadogTracing }},
+    {
+      "name": "ENVOY_STATS_CONFIG_FILE",
+      "value": "/tmp/envoy/envoyconf.yaml"
+    }{{ end }},
+    {
+      "name": "AWS_REGION",
+      "value": "{{ .Region }}"
+    }{{ if .InjectXraySidecar }},
+    {
+      "name": "ENABLE_ENVOY_XRAY_TRACING",
+      "value": "1"
+    }{{ end }}{{ if .EnableStatsTags }},
+    {
+      "name": "ENABLE_ENVOY_STATS_TAGS",
+      "value": "1"
+    }{{ end }}{{ if .EnableStatsD }},
+    {
+      "name": "ENABLE_ENVOY_DOG_STATSD",
+      "value": "1"
+    }{{ end }}
+  ]{{ if or .EnableJaegerTracing .EnableDatadogTracing }},
+  "volumeMounts": [
+    {
+      "mountPath": "/tmp/envoy",
+      "name": "envoy-tracing-config"
+    }
+  ]{{ end }},
+  "resources": {
+    "requests": {
+      "cpu": "{{ .CpuRequests }}",
+      "memory": "{{ .MemoryRequests }}"
+    }
+  }
+}
+`
+const xrayDaemonContainerTemplate = `
+{
+  "name": "xray-daemon",
+  "image": "amazon/aws-xray-daemon",
+  "securityContext": {
+    "runAsUser": 1337
+  },
+  "ports": [
+    {
+      "containerPort": 2000,
+      "name": "xray",
+      "protocol": "UDP"
+    }
+  ],
+  "env": [
+    {
+      "name": "AWS_REGION",
+      "value": "{{ .Region }}"
+    }
+  ],
+  "resources": {
+    "requests": {
+      "cpu": "{{ .CpuRequests }}",
+      "memory": "{{ .MemoryRequests }}"
+    }
+  }
+}
+`
+
+type SidecarMeta struct {
+	ContainerImage       string
+	MeshName             string
+	VirtualNodeName      string
+	Preview              string
+	LogLevel             string
+	Region               string
+	CpuRequests          string
+	MemoryRequests       string
+	EnableJaegerTracing  bool
+	JaegerAddress        string
+	JaegerPort           string
+	EnableDatadogTracing bool
+	DatadogAddress       string
+	DatadogPort          string
+	InjectXraySidecar    bool
+	EnableStatsTags      bool
+	EnableStatsD         bool
+}
+
+func renderSidecars(meta SidecarMeta) ([]string, error) {
+	var sidecars []string
```

Comment:
I have a feeling that this package is kind of randomly divided. Like why is Envoy called Sidecar, and why is Xray special and inside this `renderSidecars` function, but Jaeger and DataDog and others are in separate files? Also, the logic to configure the metadata is very spread out. For example, `JaegerAddress` is used in four different files. 

Can't we have one config and one sidecar per file, and tie them together in `config.go`? Can `SidecarMeta` have the general config as separate fields, then a slice of a `SidecarConfig` type or similar where we add all configured sidecars? Then we can just loop over those configs in `InjectSideCars()`.

I understand that there are some dependencies between them, `EnableJaegerTracing` is used in the Envoy template for example, but maybe that is something we can make more clear as well, which ones depend on others?

Reply:
Thanks for your feedback.

Envoy is always injected unless sidecar injection is disabled entirely.

Jaegar and Datadog only have the initContainers and their purpose is to inject configuration to the envoy sidecar. There are three initContainers - 
1) proxyinit (or annotations if AppMesh CNI is enabled)
2) jaeger (if configured)
3) datadog (if configured)
and Two sidecars
1) Envoy
2) X-ray (if enabled)
I ported the changes mostly as-is from the inject repository.

Based on your previous comment as well, here is what I can do - 

type SidecarRenderer interface {
   renderInitContainer() (corev1.Container, error)
   renderContainer() (corev1.Container, error)
   renderConfigVolume() (corev1.Volume, error)
}
Each sidecar file will implement the interface, will return nil objects if not configured of not applicable. Then have the iterator in InjectSideCars() add the valid objects returned by the modules.